### DISCUSSION TITLE
How to test UI changes?

### DISCUSSION THREAD

**Discussion by lilfrootyloops** [cytoscape/cytoscape.js#3208] [Q&A] [answered]
Are there any tools we can use to test changes in the `<canvas>` that are caused by Cytoscape.js? I've tried searching for any existing libraries that may do so, but it doesn't appear that there are any.

**Comment by maxkfranz** [comment, marked as answer]
Playwright, or any other general tool that allows for pixel-perfect, screenshot-based testing: 

- https://playwright.dev
- https://duckduckgo.com/?q=pixel+perfect+browser+ui+testing

**Reply by lilfrootyloops** [reply]
Thanks for the suggestion & letting me know about pixel-perfect, screenshot-based testing.
